### PR TITLE
B state machine

### PR DESCRIPTION
**Pull request by brianyenna**
This PR implements the following features:
1. Add GameStateMachine (and the relevant states)
2. Add GameEndScene and GamePauseScene
3. Add SceneManager

GameStateMachine manages the transitions between the various states. The idea behind it is to expose an interface so that whenever we code out the game, we simply need to make GameStateMachine enter a different state, and GameStateMachine will handle the logic required to handle the state transition (i.e. transition to a different scene, pause enemy movement etc.)

SceneManager manages the actual transitions between different scenes (not states, that's done by GameStateMachine). 

This PR also implements the decrementing of health from Model. Once the health points drop to zero, the game transitions to the end (lose) state. However, this feature is incomplete. Pending Andy's monster-endpoint collision code and monster-removal code, and pending Jeremy's UI-Model linking code.

### CODE REVIEW

File: GameOfRunes/GameOfRunes/Custom Nodes/ButtonNode.swift [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+//
+
+import SpriteKit
+
+/**
+ A type that can respond to `ButtonNode` button press events.
+ This should be implemented by the `SKScene` that the `ButtonNode`
+ is added to.
+ */
+protocol ButtonNodeResponderType : class {
```
Comment (dongsiji) [inline GameOfRunes/GameOfRunes/Custom Nodes/ButtonNode.swift]: Please refactor to a separate file

File: GameOfRunes/GameOfRunes/Entities/EntityManager.swift [hunk 2]
```diff
@@ -51,6 +57,16 @@ class EntityManager {
             }
         }
         toRemoveEntities = []
+
+        // Player Loses the Game
+        if let playerHealthEntity = playerHealthEntity,
+            let playerHealthComponent = playerHealthEntity.component(ofType: HealthComponent.self),
+            playerHealthComponent.healthPoints == 0,
```
Comment (dongsiji) [inline GameOfRunes/GameOfRunes/Entities/EntityManager.swift]: use <= 0

File: GameOfRunes/GameOfRunes/Custom Nodes/ButtonNode.swift (hunk 1 above)
Comment (brianyenna) [reply GameOfRunes/GameOfRunes/Custom Nodes/ButtonNode.swift]: Done! Resolved in commit 50cadfd

File: GameOfRunes/GameOfRunes/Entities/EntityManager.swift (hunk 2 above)
Comment (brianyenna) [reply GameOfRunes/GameOfRunes/Entities/EntityManager.swift]: Resolved in commit 50cadfd

**Review by dongsiji — APPROVED** [review]

**Pull request merged**